### ISSUE TITLE
Function `torch.median` doesn't release memory after usage on GPU

### ISSUE THREAD

**Issue by tgadeliya** [pytorch/pytorch#112895]
### 🐛 Describe the bug

When I am using `torch.median` on GPU I experiencing memory leak. I am not sure whether it is expected behaviour, but used memory on GPU increasing x3-4 times and not releasing automatically. On `cpu` memory is releasing automatically
Below example replicates this problem 
```python
import torch

# I am using this gpustat and ipdb to report used GPU memory
# GPU memory : 449 mb
a = torch.empty(32000, 768).to("cuda") # GPU memory : 1863 mb
torch.median(a) # GPU memory : 9453 mb (and remains on this level)
```
Triggering `torch.cuda.empty_cache()` after `torch.median` call releases GPU memory
The same behaviour I am observing on Colab

### Expected behaviour:
 ```
Memory is automatically released after `torch.median` call for tensor on GPU
```

### Versions

Collecting environment information...
PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.27.7
Libc version: glibc-2.31

Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.10.0-25-cloud-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             8
On-line CPU(s) list:                0-7
Thread(s) per core:                 2
Core(s) per socket:                 4
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              79
Model name:                         Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                           0
CPU MHz:                            2199.998
BogoMIPS:                           4399.99
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          128 KiB
L1i cache:                          128 KiB
L2 cache:                           1 MiB
L3 cache:                           55 MiB
NUMA node0 CPU(s):                  0-7
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT Host state unknown
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.1
[pip3] torch==2.0.0
[pip3] triton==2.0.0
[conda] dlenv-pytorch-1-13-gpu    1.0.20230926    py310h003b471_0    file:///tmp/conda-pkgs
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] torch                     1.13.1                   pypi_0    pypi
[conda] torchvision               0.14.1                   pypi_0    pypi

**Comment by tringwald** [comment]
I can't reproduce this on `torch==2.0.0+cu117` (or any newer version). With the given code snippet, `torch` allocates ~1GB of GPU memory and I didn't observe any increase in memory usage afterwards. 
As you mentioned Colab: Do you also observe this problem when not running the snipped inside of a Python notebook?

**Comment by tgadeliya** [comment]
> As you mentioned Colab: Do you also observe this problem when not running the snippet inside of a Python notebook?
 
 Yes, This problem was initially found in python script, when running normally (not notebook)


I used [this](https://colab.research.google.com/drive/1BWsM4RaaAG9sRfpYIhacdnxiwRe-uaoa?usp=sharing) Colab only for reproducing problem in other environment and other version.  You can to run this colab on T4 (for example) and see spikes and memory  in GUI:

after tensor created:
<img width="161" alt="image" src="https://github.com/pytorch/pytorch/assets/32731151/0800b934-5c5a-4880-8d04-2574d93286c7">

after torch.median is called:
<img width="167" alt="image" src="https://github.com/pytorch/pytorch/assets/32731151/0915e749-15cb-466a-8b20-c279ee7d40f1">

**Comment by tringwald** [comment]
With the code from Colab, I am actually able to reproduce the problem (also on `2.1.0` and different GPUs). However, the code is slightly different from the snippet posted above. 

```python3
import torch

with torch.no_grad():
    # Note the 320_000 instead of 32_000
    a = torch.ones(320000, 768, requires_grad=False).cuda()
    b = torch.median(a)
# Endless loop, so we can observe memory usage in nvidia-smi
while True:
    pass
```

Without the `torch.median` line, this consumes ~1.2GB of VRAM; with the line included it's ~8.5GB.

**Comment by tgadeliya** [comment]
> # Note the 320_000 instead of 32_000
Sorry, missed zero I think. But nevertheless, this problem will appear and with 32_000 tensor size. The only different size is GPU memory spike size  

> # Endless loop, so we can observe memory usage in nvidia-smi
As I wrote above, I located this problem using `gpustat` and `ipdb`. Using normal run you will also have this problem, but it will be difficult to spot in such small script. But if you set tensor size large enough to fit 80-90% of GPU and run `torch.median` I you will have OOM
